File: qqtech9-backend/src/repositories/runners.js
import { pool } from '../config/database-conn.js';
import { Runners, RunnerQueue, RunnerLogs } from '../models/runners.js';

export const RunnersRepository = {
    findAllPaginatedWithFilters: async (
        ruleName, status, priority, databaseType, isSuperAdmin, roles, limit, offset
    ) => {
        const selectQuery = 
        `
        SELECT 
            r.*,
            json_build_object(
                'id', rl.id,
                'name', rl.name,
                'database_type', rl.database_type,
                'priority', rl.priority,
                'execution_interval_ms', rl.execution_interval_ms,
                'is_active', rl.is_active
            ) AS rule
        FROM runners r
        LEFT JOIN rules rl 
            ON r.rule_id = rl.id
        WHERE
            ($1::varchar IS NULL OR rl.name ILIKE '%' || $1 || '%')
            AND ($2::varchar IS NULL OR r.status = $2)
            AND ($3::varchar IS NULL OR rl.priority = $3)
            AND ($4::varchar IS NULL OR rl.database_type = $4)
            AND (
                $5::boolean = true
                OR (
                    $6::uuid[] IS NOT NULL
                    AND EXISTS (
                        SELECT 1
                        FROM rules_roles rr_auth
                        WHERE rr_auth.rule_id = rl.id
                        AND rr_auth.role_id = ANY($6::uuid[])
                    )
                )
            )
        GROUP BY r.id, rl.id
        ORDER BY last_run_at DESC
        LIMIT $7 OFFSET $8;
        `;


        const values = [
            ruleName || null,
            status || null,
            priority || null,
            databaseType || null,
            isSuperAdmin,
            roles?.length ? roles : null,
            limit,
            offset,
        ];

        const result = await pool.query(selectQuery, values);

        return Runners.fromArray(result.rows);
    },

    findById: async (id) => {
        const selectIdQuery = 
        `
        SELECT * FROM runners
        WHERE id = $1
        `;

        const result = await pool.query(selectIdQuery, [id]);

        if(!result.rows[0]){
            return null;
        }

        return new Runners(result.rows[0]);
    },

    create: async (runner, client = pool) => {
        const insertRunnerQuery = 
        `
        INSERT INTO runners
        (rule_id, status, last_run_at)
        VALUES ($1, $2, $3)
        RETURNING *; 
        `;

        const values = [
            runner.ruleId,
            runner.status,
            runner.lastRunAt,
        ];

        const result =  await client.query(insertRunnerQuery, values);
        
        return new Runners(result.rows[0]);
    },

    update: async (runner) => {
        const updateQuery = 
        `
        UPDATE runners
        SET rule_id = $1,
            status = $2,
            last_run_at = $3
        WHERE id = $4
        RETURNING *;
        `;

        const values = [
            runner.ruleId,
            runner.status,
            runner.lastRunAt,
            runner.id
        ];

        const result = await pool.query(updateQuery, values);

        return new Runners(result.rows[0]);
    },

    delete: async (id) => {
        const deleteQuery = 
        `
        DELETE FROM runners
        WHERE id = $1;
        `;

        await pool.query(deleteQuery, [id]);
    }
};

export const RunnerQueueRepository = {
    findAll: async (
        ruleName, status, rulePriority, isSuperAdmin, roles, limit, offset
    ) => {
        const selectQuery =
        `
        SELECT 
            rq.*,
            json_build_object(
                'id', rl.id,
                'name', rl.name,
                'priority', rl.priority
            
            ) AS rule
        FROM runner_queue rq
        LEFT JOIN runners r 
            ON rq.runner_id = r.id
        LEFT JOIN rules rl
            ON r.rule_id = rl.id
        WHERE
            ($1::varchar IS NULL OR rl.name ILIKE '%' || $1 || '%')
            AND ($2::varchar IS NULL OR rq.status = $2)
            AND ($3::varchar IS NULL OR rl.priority = $3)
            AND (
                $4::boolean = true
                OR (
                    $5::uuid[] IS NOT NULL
                    AND EXISTS (
                        SELECT 1
                        FROM rules_roles rr_auth
                        WHERE rr_auth.rule_id = rl.id
                        AND rr_auth.role_id = ANY($5::uuid[])
                    )
                )
            )
        GROUP BY rq.id, rl.id
        ORDER BY scheduled_for DESC
        LIMIT $6 OFFSET $7;
        `;

        const value = [
            ruleName || null,
            status || null,
            rulePriority || null,
            isSuperAdmin,
            roles?.length ? roles : null,
            limit,
            offset,
        ];

        const result = await pool.query(selectQuery, value);

        return RunnerQueue.fromArray(result.rows);
    },

    findById: async (id) => {
        const selectIdQuery = 
        `
        SELECT * FROM runner_queue
        WHERE id = $1
        `;

        const result = await pool.query(selectIdQuery, [id]);

        if(!result.rows[0]){
            return null;
        }

        return new RunnerQueue(result.rows[0]);
    },

    create: async (runnerQueue) => {
        const insertQuery = 
        `
        INSERT INTO runner_queue
        (runner_id, scheduled_for)
        VALUES ($1, $2)
        RETURNING *;
        `

        const values = [
            runnerQueue.runnerId,
            runnerQueue.scheduledFor,
        ];

        const result = await pool.query(insertQuery, values);

        return new RunnerQueue(result.rows[0]);
    },

    update: async (runnerQueue) => {
        const updateQuery = 
        `
        UPDATE runner_queue
        SET status = $1,
            scheduled_for = $2,
            started_at = $3,
            finished_at = $4,
            attempt_count = $5
        WHERE id = $6
        RETURNING *;
        `;

        const values = [
            runnerQueue.status,
            runnerQueue.scheduledFor,
            runnerQueue.startedAt,
            runnerQueue.finishedAt,
            runnerQueue.attemptCount,
            runnerQueue.id
        ];

        const result = await pool.query(updateQuery, values);

        return new RunnerQueue(result.rows[0]);
    },

    delete: async (id) => {
        const deleteQuery = 
        `
        DELETE FROM runner_queue
        WHERE id = $1;
        `;

        await pool.query(deleteQuery, [id]);
    }
};

export const RunnerLogsRepository = {
    findAll: async (
        ruleName, executionStatus, limit, offset
    ) => {
        const selectQuery =
        `
        SELECT 
            rlogs.*,
            json_build_object(
                'id', rl.id,
                'name', rl.name,
                'database_type', rl.database_type,
                'priority', rl.priority
            ) AS rule
        FROM runner_logs rlogs
        LEFT JOIN runners r
            ON rlogs.runner_id = r.id
        LEFT JOIN rules rl
            ON r.rule_id = rl.id
        WHERE
            ($1::varchar IS NULL OR rl.name ILIKE '%' || $1 || '%')
            AND ($2::varchar IS NULL OR rlogs.execution_status = $2)
        GROUP BY rlogs.id, rl.id
        ORDER BY executed_at DESC
        LIMIT $3 OFFSET $4;
        `;

        const values = [
            ruleName || null,
            executionStatus || null,
            limit,
            offset,
        ];

        const result = await pool.query(selectQuery, values);


        return RunnerLogs.fromArray(result.rows);
    },

    create: async (runnerLog) => {
        const insertQuery = 
        `
        INSERT INTO runner_logs
        (runner_id, queue_id, run_time_ms, execution_status, rows_affected, result, error, executed_at)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
        RETURNING *;
        `

        const values = [
            runnerLog.runnerId,
            runnerLog.queueId,
            runnerLog.runTimeMs,
            runnerLog.executionStatus,
            runnerLog.rowsAffected, 
            runnerLog.result,
            runnerLog.error,
            runnerLog.executedAt
        ];

        const result = await pool.query(insertQuery, values);

        return new RunnerLogs(result.rows[0]);
    }
}

File: python-workers/repositories/runners.py
from sqlalchemy import  text
from config.database import get_session
from models.runners import Runners, RunnerQueue, RunnerLogs


class RunnersRepository:
    @staticmethod
    async def find_all_for_scheduling():
        query = text(
        """
        SELECT r.*
        FROM runners r
        LEFT JOIN rules rl
            ON r.rule_id = rl.id
        ORDER BY 
            CASE rl.priority
                WHEN 'HIGH' THEN 1
                WHEN 'MEDIUM' THEN 2
                WHEN 'LOW' THEN 3
                ELSE 4
            END;
        """
        )
        
        async with get_session() as session:
            result = await session.execute(query)
            rows = result.fetchall()
            return Runners.from_array([dict(row._mapping) for row in rows])

    @staticmethod
    async def find_by_id(runner_id):
        async with get_session() as session:
            query = text( 
            """
            SELECT * FROM runners
            WHERE id = :id
            """
            ).bindparams(id=runner_id)
            
            result = await session.execute(query)
            row = result.fetchone()
            
            if not row:
                return None
            
            return Runners(**dict(row._mapping))

    @staticmethod
    async def update(runner):
        async with get_session() as session:
            query=text(
            """
            UPDATE runners
            SET rule_id = :rule_id,
                status = :status,
                last_run_at = :last_run_at
            WHERE id = :id
            RETURNING *;
            """
            )
            
            query = query.bindparams(
                rule_id = runner.rule_id,
                status = runner.status,
                last_run_at = runner.last_run_at,
                id = runner.id
            )
            
            result = await session.execute(query)
            await session.commit()
            row = result.fetchone()
            return Runners(**dict(row._mapping))


class RunnerQueueRepository:
    @staticmethod
    async def find_pending_jobs(limit):
        query = text(
        """
        SELECT * FROM runner_queue
        WHERE status = 'PENDING'
        AND scheduled_for AT TIME ZONE 'UTC' <= NOW()
        ORDER BY scheduled_for ASC
        LIMIT :limit;
        """
        ).bindparams(limit=limit)
        
        async with get_session() as session:
            result = await session.execute(query)
            rows = result.fetchall()
            return RunnerQueue.from_array([dict(row._mapping) for row in rows])

    @staticmethod
    async def find_pending_by_runner_id(runner_id):
        query = text(
        """
        SELECT * FROM runner_queue
        WHERE runner_id = :runner_id
        AND status in ('PENDING', 'PROCESSING')
        LIMIT 1;
        """
        ).bindparams(runner_id=runner_id)
        
        async with get_session() as session:
            result = await session.execute(query)
            row = result.fetchone()
            
            if not row:
                return None
            
            return RunnerQueue(**dict(row._mapping))

    @staticmethod
    async def create(runner_queue):
        query = text(
        """
        INSERT INTO runner_queue
        (runner_id, scheduled_for)
        VALUES (:runner_id, :scheduled_for)
        RETURNING *;
        """
        )
        
        query = query.bindparams(
            runner_id=runner_queue.runner_id,
            scheduled_for=runner_queue.scheduled_for
        )
        
        
        async with get_session() as session:
            result = await session.execute(query)
            await session.commit()
            row = result.fetchone()
            return RunnerQueue(**dict(row._mapping))

    @staticmethod
    async def update(runner_queue):
        query = text(
        """
        UPDATE runner_queue
        SET status = :status,
            scheduled_for = :scheduled_for,
            queued_at = :queued_at,
            started_at = :started_at,
            finished_at = :finished_at,
            attempt_count = :attempt_count
        WHERE id = :id
        RETURNING *;
        """
        )
        
        query = query.bindparams(
            status=runner_queue.status,
            scheduled_for=runner_queue.scheduled_for,
            queued_at=runner_queue.queued_at,
            started_at=runner_queue.started_at,
            finished_at=runner_queue.finished_at,
            attempt_count=runner_queue.attempt_count,
            id=runner_queue.id
        )
        
        async with get_session() as session:
            result = await session.execute(query)   
            await session.commit()
            row = result.fetchone()
            return RunnerQueue(**dict(row._mapping))


class RunnerLogsRepository:
    @staticmethod
    async def create(runner_log):
        query = text(
        """
        INSERT INTO runner_logs
        (runner_id, queue_id, run_time_ms, execution_status, rows_affected, result, error, executed_at)
        VALUES (:runner_id, :queue_id, :run_time_ms, :execution_status, :rows_affected, :result, :error, :executed_at)
        RETURNING *;
        """
        )
        
        query = query.bindparams(
            runner_id=runner_log.runner_id,
            queue_id=runner_log.queue_id,
            run_time_ms=runner_log.run_time_ms,
            execution_status=runner_log.execution_status,
            rows_affected=runner_log.rows_affected,
            result=runner_log.result,
            error=runner_log.error,
            executed_at=runner_log.executed_at
        )
        
        async with get_session() as session:
            result = await session.execute(query)
            await session.commit()
            row = result.fetchone()
            return RunnerLogs(**dict(row._mapping))
